tests/test_reply_function_sync.js

```diff
@@ -48,152 +33,199 @@ test('get with reply callback returning object', async t => {
+// `.reply(() => [200, 'ABC', { 'X-My-Headers': 'My custom header value' }])`
+test('reply with status code, function returning string body, and header object', async t => {
+  const scope = nock('http://example.com')
+    .get('/')
+    .reply(200, () => 'ABC', { 'X-My-Headers': 'My custom header value' })
+
+  const { headers } = await got('http://example.com/')
+
+  t.equivalent(headers, { 'x-my-headers': 'My custom header value' })
 
-  t.equal(body, 'overwrite')
   scope.done()
 })
 
-test('reply can take a callback', async t => {
+test(
+  'reply function returning array with status code',
+  // Seems likely a bug related to https://github.com/nock/nock/issues/1222.
+  { skip: true },
+  async t => {
+    const scope = nock('http://example.test')
+      .get('/')
+      .reply(() => [202])
+
+    const { statusCode, body } = await got('http://example.test/')
+
+    t.is(statusCode, 202)
+    t.equal(body, '')
+    scope.done()
+  }
+)
+
+test('reply function returning array with status code and string body', async t => {
   const scope = nock('http://example.com')
     .get('/')
-    .reply(200, (path, requestBody, callback) => callback(null, 'Hello World!'))
+    .reply(() => [401, 'This is a body'])
 
-  const response = await got('http://example.com/', {
-    encoding: null,
+  await assertRejects(got('http://example.com/'), err => {
+    t.equal(err.statusCode, 401)
+    t.equal(err.body, 'This is a body')
+    return true
   })
+  scope.done()
+})
+
+test('reply function returning array with status code and body object', async t => {
+  const exampleResponse = { message: 'OK!' }
+
+  const scope = nock('http://example.test')
+    .get('/')
+    .reply(() => [202, exampleResponse])
+
+  const { statusCode, body } = await got('http://example.test/')
 
+  t.is(statusCode, 202)
+  t.equal(body, JSON.stringify(exampleResponse))
   scope.done()
-  t.type(response.body, Buffer)
-  t.equal(response.body.toString('utf8'), 'Hello World!')
 })
 
-test('reply should send correct statusCode with array-notation and without body', async t => {
-  t.plan(1)
+test('reply function returning array with status code and body as number', async t => {
+  const scope = nock('http://example.test')
+    .get('/')
+    .reply(() => [202, 123])
 
-  const expectedStatusCode = 202
+  const { statusCode, body } = await got('http://example.test/')
 
-  const scope = nock('http://example.com')
+  t.is(statusCode, 202)
+  t.type(body, 'string')
+  t.equal(body, '123')
+  scope.done()
+})
+
+test('reply function returning array with status code, string body, and headers object', async t => {
+  const scope = nock('http://example.test')
     .get('/')
-    .reply((path, requestBody) => [expectedStatusCode])
+    .reply(() => [202, 'body', { 'x-key': 'value', 'x-key-2': 'value 2' }])
 
-  const { statusCode } = await got('http://example.com/')
+  const { headers, rawHeaders } = await got('http://example.test/')
 
-  t.equal(statusCode, expectedStatusCode)
```

Comment:
With the amount of hand waving that happens around optional data, it's probably worth explicitly asserting the status code and body down in these tests.

Reply:
Good shout. Updated!